File: bubus-ts/src/bridge_redis.ts
/**
 * Redis pub/sub bridge for forwarding events between runtimes.
 *
 * Usage:
 *   // channel from URL path
 *   const bridge = new RedisEventBridge('redis://user:pass@localhost:6379/1/my_channel')
 *
 *   // explicit channel override
 *   const bridge2 = new RedisEventBridge('redis://user:pass@localhost:6379/1', 'my_channel')
 *
 * URL format:
 *   redis://user:pass@host:6379/<db>/<optional_channel>
 */
import { BaseEvent } from './base_event.js'
import { EventBus } from './event_bus.js'
import { assertOptionalDependencyAvailable, importOptionalDependency, isNodeRuntime } from './optional_deps.js'
import type { EventClass, EventHandlerFunction, EventKey, UntypedEventHandlerFunction } from './types.js'

const randomSuffix = (): string => Math.random().toString(36).slice(2, 10)
const DEFAULT_REDIS_CHANNEL = 'bubus_events'
const DB_INIT_KEY = '__bubus:bridge_init__'

const parseRedisUrl = (redis_url: string, channel?: string): { url: string; channel: string } => {
  let parsed: URL
  try {
    parsed = new URL(redis_url)
  } catch {
    throw new Error(`RedisEventBridge URL must be a valid redis:// or rediss:// URL, got: ${redis_url}`)
  }

  const protocol = parsed.protocol.replace(/:$/, '').toLowerCase()
  if (protocol !== 'redis' && protocol !== 'rediss') {
    throw new Error(`RedisEventBridge URL must use redis:// or rediss://, got: ${redis_url}`)
  }

  const segments = parsed.pathname.split('/').filter(Boolean)
  if (segments.length > 2) {
    throw new Error(`RedisEventBridge URL path must be /<db> or /<db>/<channel>, got: ${parsed.pathname || '/'}`)
  }

  let db_index = '0'
  let channel_from_url: string | undefined

  if (segments.length > 0) {
    db_index = segments[0]
    if (!/^\d+$/.test(db_index)) {
      throw new Error(`RedisEventBridge URL db path segment must be numeric, got: ${JSON.stringify(db_index)} in ${redis_url}`)
    }
    if (segments.length === 2) {
      channel_from_url = segments[1]
    }
  }

  const resolved_channel = channel ?? channel_from_url ?? DEFAULT_REDIS_CHANNEL
  if (!resolved_channel) {
    throw new Error('RedisEventBridge channel must not be empty')
  }

  const normalized = new URL(parsed.toString())
  normalized.pathname = `/${db_index}`
  return { url: normalized.toString(), channel: resolved_channel }
}

export class RedisEventBridge {
  readonly url: string
  readonly channel: string
  readonly name: string

  private readonly inbound_bus: EventBus
  private running: boolean
  private start_promise: Promise<void> | null
  private redis_pub: any | null
  private redis_sub: any | null

  constructor(redis_url: string, channel?: string, name?: string) {
    assertOptionalDependencyAvailable('RedisEventBridge', 'ioredis')

    const parsed = parseRedisUrl(redis_url, channel)
    this.url = parsed.url
    this.channel = parsed.channel
    this.name = name ?? `RedisEventBridge_${randomSuffix()}`
    this.inbound_bus = new EventBus(this.name, { max_history_size: 0 })
    this.running = false
    this.start_promise = null
    this.redis_pub = null
    this.redis_sub = null

    this.dispatch = this.dispatch.bind(this)
    this.emit = this.emit.bind(this)
    this.on = this.on.bind(this)
  }

  on<T extends BaseEvent>(event_key: EventClass<T>, handler: EventHandlerFunction<T>): void
  on<T extends BaseEvent>(event_key: string | '*', handler: UntypedEventHandlerFunction<T>): void
  on(event_key: EventKey | '*', handler: EventHandlerFunction | UntypedEventHandlerFunction): void {
    this.ensureStarted()
    if (typeof event_key === 'string') {
      this.inbound_bus.on(event_key, handler as UntypedEventHandlerFunction<BaseEvent>)
      return
    }
    this.inbound_bus.on(event_key as EventClass<BaseEvent>, handler as EventHandlerFunction<BaseEvent>)
  }

  async dispatch<T extends BaseEvent>(event: T): Promise<void> {
    this.ensureStarted()
    if (!this.redis_pub) await this.start()
    const payload = JSON.stringify(event.toJSON())
    await this.redis_pub.publish(this.channel, payload)
  }

  async emit<T extends BaseEvent>(event: T): Promise<void> {
    return this.dispatch(event)
  }

  async start(): Promise<void> {
    if (this.running) return
    if (this.start_promise) {
      await this.start_promise
      return
    }

    // `on(...)` auto-start and explicit `await start()` can happen back-to-back; use one in-flight
    // startup promise so we do not leak extra Redis clients.
    this.start_promise = (async () => {
      if (!isNodeRuntime()) {
        throw new Error('RedisEventBridge is only supported in Node.js runtimes')
      }

      const mod = await importOptionalDependency('RedisEventBridge', 'ioredis')
      const Redis = mod.default ?? mod.Redis ?? mod
      const redis_pub = new Redis(this.url)
      const redis_sub = new Redis(this.url)

      redis_pub.on('error', () => {})
      redis_sub.on('error', () => {})

      // Redis logical DBs are created lazily; writing a short-lived key initializes/validates the selected DB.
      await redis_pub.set(DB_INIT_KEY, '1', 'EX', 60, 'NX')
      redis_sub.on('message', (channel_name: string, message: string) => {
        if (channel_name !== this.channel) return
        try {
          const payload = JSON.parse(message)
          void this.dispatchInboundPayload(payload)
        } catch {
          // Ignore malformed payloads.
        }
      })
      await redis_sub.subscribe(this.channel)
      this.redis_pub = redis_pub
      this.redis_sub = redis_sub
      this.running = true
    })()

    try {
      await this.start_promise
    } finally {
      this.start_promise = null
    }
  }

  async close(): Promise<void> {
    if (this.start_promise) {
      await this.start_promise.catch(() => {})
    }
    this.running = false
    if (this.redis_sub) {
      try {
        await this.redis_sub.unsubscribe(this.channel)
      } catch {
        // ignore
      }
      await this.redis_sub.quit()
      this.redis_sub = null
    }
    if (this.redis_pub) {
      await this.redis_pub.quit()
      this.redis_pub = null
    }
    this.inbound_bus.destroy()
  }

  private ensureStarted(): void {
    if (this.running) return
    if (this.start_promise) return
    void this.start().catch((error: unknown) => {
      console.error('[bubus] RedisEventBridge failed to start', error)
    })
  }

  private async dispatchInboundPayload(payload: unknown): Promise<void> {
    const parsed_event = BaseEvent.fromJSON(payload)
    const existing_event = EventBus._all_instances.findEventById(parsed_event.event_id)
    const event = existing_event ?? parsed_event.reset()
    this.inbound_bus.dispatch(event)
  }
}

File: bubus/bridge_redis.py
"""Redis pub/sub bridge for forwarding events between runtimes.

Optional dependency: redis>=5 (uses redis.asyncio)

Usage:
    # channel from URL path
    bridge = RedisEventBridge('redis://user:pass@localhost:6379/1/my_channel')

    # explicit channel override
    bridge = RedisEventBridge('redis://user:pass@localhost:6379/1', channel='my_channel')

Connection URL format:
    redis://user:pass@host:6379/1/channel_name

The first path segment is the Redis logical DB index.
An optional second path segment is used as the pub/sub channel.
If channel is omitted in both URL and constructor, defaults to "bubus_events".
"""

from __future__ import annotations

import asyncio
import importlib
import json
from collections.abc import Callable
from typing import Any, cast
from urllib.parse import urlsplit, urlunsplit

from uuid_extensions import uuid7str

from bubus.models import BaseEvent
from bubus.service import EventBus, EventPatternType, inside_handler_context

_DEFAULT_REDIS_CHANNEL = 'bubus_events'
_DB_INIT_KEY = '__bubus:bridge_init__'


def _parse_redis_url(redis_url: str, channel: str | None) -> tuple[str, str]:
    parsed = urlsplit(redis_url)
    scheme = parsed.scheme.lower()
    if scheme not in ('redis', 'rediss'):
        raise ValueError(f'RedisEventBridge URL must use redis:// or rediss://, got: {redis_url}')

    path_segments = [segment for segment in parsed.path.split('/') if segment]
    if len(path_segments) > 2:
        raise ValueError(f'RedisEventBridge URL path must be /<db> or /<db>/<channel>, got: {parsed.path or "/"}')

    db_index = '0'
    channel_from_url: str | None = None
    if path_segments:
        db_index = path_segments[0]
        if not db_index.isdigit():
            raise ValueError(f'RedisEventBridge URL db path segment must be numeric, got: {db_index!r} in {redis_url}')
        if len(path_segments) == 2:
            channel_from_url = path_segments[1]

    resolved_channel = channel or channel_from_url or _DEFAULT_REDIS_CHANNEL
    if not resolved_channel:
        raise ValueError('RedisEventBridge channel must not be empty')

    normalized_path = f'/{db_index}'
    normalized_url = urlunsplit((parsed.scheme, parsed.netloc, normalized_path, parsed.query, parsed.fragment))
    return normalized_url, resolved_channel


class RedisEventBridge:
    def __init__(self, redis_url: str, channel: str | None = None, *, name: str | None = None):
        self.url, self.channel = _parse_redis_url(redis_url, channel)
        self._inbound_bus = EventBus(name=name or f'RedisEventBridge_{uuid7str()[-8:]}', max_history_size=0)

        self._running = False
        self._listener_task: asyncio.Task[None] | None = None
        self._redis_pub: Any | None = None
        self._redis_sub: Any | None = None
        self._pubsub: Any | None = None

    def on(self, event_pattern: EventPatternType, handler: Callable[[BaseEvent[Any]], Any]) -> None:
        self._ensure_started()
        self._inbound_bus.on(event_pattern, handler)

    async def dispatch(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        self._ensure_started()
        if self._redis_pub is None:
            await self.start()

        payload = event.model_dump(mode='json')
        assert self._redis_pub is not None
        await self._redis_pub.publish(self.channel, json.dumps(payload, separators=(',', ':')))

        if inside_handler_context.get():
            return None
        return event

    async def emit(self, event: BaseEvent[Any]) -> BaseEvent[Any] | None:
        return await self.dispatch(event)

    async def start(self) -> None:
        if self._running:
            return

        redis_asyncio = self._load_redis_asyncio()
        self._redis_pub = redis_asyncio.from_url(self.url, decode_responses=True)
        self._redis_sub = redis_asyncio.from_url(self.url, decode_responses=True)
        assert self._redis_pub is not None
        assert self._redis_sub is not None

        # Redis logical DBs are created lazily; writing a short-lived key initializes/validates the selected DB.
        await self._redis_pub.set(_DB_INIT_KEY, '1', ex=60, nx=True)

        self._pubsub = self._redis_sub.pubsub()
        assert self._pubsub is not None
        await self._pubsub.subscribe(self.channel)

        self._running = True
        self._listener_task = asyncio.create_task(self._listen_loop())

    async def close(self, *, clear: bool = True) -> None:
        self._running = False
        if self._listener_task is not None:
            self._listener_task.cancel()
            await asyncio.gather(self._listener_task, return_exceptions=True)
            self._listener_task = None

        if self._pubsub is not None:
            await self._pubsub.unsubscribe(self.channel)
            await self._pubsub.close()
            self._pubsub = None
        if self._redis_sub is not None:
            await self._redis_sub.close()
            self._redis_sub = None
        if self._redis_pub is not None:
            await self._redis_pub.close()
            self._redis_pub = None

        await self._inbound_bus.stop(clear=clear)

    def _ensure_started(self) -> None:
        if self._running:
            return
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return
        self._listener_task = asyncio.create_task(self.start())

    async def _listen_loop(self) -> None:
        assert self._pubsub is not None
        try:
            async for message in self._pubsub.listen():
                if not self._running:
                    break
                if not isinstance(message, dict):
                    continue
                message_dict = cast(dict[str, Any], message)
                if message_dict.get('type') != 'message':
                    continue

                raw_data = message_dict.get('data')
                if isinstance(raw_data, bytes):
                    data = raw_data.decode('utf-8')
                elif isinstance(raw_data, str):
                    data = raw_data
                else:
                    continue

                try:
                    payload = json.loads(data)
                except Exception:
                    continue
                await self._dispatch_inbound_payload(payload)
        except asyncio.CancelledError:
            raise
        except Exception:
            if self._running:
                await asyncio.sleep(0.05)

    async def _dispatch_inbound_payload(self, payload: Any) -> None:
        event = BaseEvent[Any].model_validate(payload).reset()
        self._inbound_bus.dispatch(event)

    @staticmethod
    def _load_redis_asyncio() -> Any:
        try:
            return importlib.import_module('redis.asyncio')
        except ModuleNotFoundError as exc:
            raise RuntimeError('RedisEventBridge requires optional dependency: pip install redis') from exc
